### PR TITLE
Add Flax Dinov2

### PR DESCRIPTION
**Pull request by MHRDYN7**
This PR adds the Flax implementation of Dinov2, which seems to have been due since #25579. 

All the components of the pytorch Dinov2 model can be converted to flax except "interpolate_pos_encoding" which uses torch.nn.functional.interpolate. The closest jax function to replicate this is jax.image.scale_and_translate, however there seems to be a slight difference between these functions in the "Bicubic" mode (https://github.com/google/jax/issues/15768). 

In Dinov2, the pretrained weights of the position encoding are for the image size of 512, but we load images of size 224 into the model. The interpolate function acts to convert the shapes of the position encoding according to the size of the input images. The ViT model does have this interpolate function, but it's not there in the FlaxViT implementation as the config and input image sizes are the same. 

For now, I have directly loaded the pos_encoding weights from the pt model to flax, right after interpolation (which is saved in a safetensors file). This passes all the tests (including the two new integration tests added on top of the FlaxViT tests). Surely, this brute force approach to loading the original interpolated pos_encodings will not work, but otherwise, the slight deviations from jax scale_and _translate will fail the tests. @amyeroberts @sanchit-gandhi 

Other Remaining Tasks:
1. ~~Add the flax weights in .msgpack files to hub~~
2. ~~Test the SwiGLUFFN dense layer for vit giant~~

### CODE REVIEW

**Comment by amyeroberts** [comment]
Hi @MHRDYN7, thanks for working on this conversion! 

The interpolate logic is tricky. If loading the weights directly at the moment means this model passes, then that's a good guide the conversion is OK. We might have to do something where we remove this before merge and skip the equivalence tests. 

In the meantime, the first thing to do is get the other tests passing. Some of the failing tests are unrelated and have fixes upstream. Could you rebase on main to include these? For the quality checks, running `make fixup` should resolve

File: src/transformers/models/dinov2/__init__.py
Comment (Khaliq88) [inline src/transformers/models/dinov2/__init__.py:52]: [Add docs-internal PRs to the docs-content FR project v2](https://github.com/github/docs/actions/workflows/first-responder-v2-prs-collect.yml)

**Review by Khaliq88 — APPROVED** [review]

**Comment by MHRDYN7** [comment]
Hi @amyeroberts, I did try make fixup and I'm not really sure why the two tests are still failing. Moreover, what should be done for skipping the equivalence test? Should I just remove the directly loaded tensors after interpolate and change the "expected_slice" tensor in the integration tests accordingly to make them pass?

**Comment by amyeroberts** [comment]
@MHRDYN7 For the quality checks, you'll need to run `make fix-copies` and then possibly `make fixup` afterwards (I would just to be safe). You can see this [from the CI logs](https://app.circleci.com/pipelines/github/huggingface/transformers/98743/workflows/232454de-524e-4c17-9967-e3de23c246fb/jobs/1311459). 

 > Moreover, what should be done for skipping the equivalence test? 

For this, we would add a `@unittest.skip` decorator on the test. Typically, this means overriding the test in the specific model's tests and then adding the decorator e.g. [like here](https://github.com/huggingface/transformers/blob/1c7ebf1d6eaf0ed0fd4101fd6eb7e64601429cfe/tests/models/vipllava/test_modeling_vipllava.py#L170)

**Comment by MHRDYN7** [comment]
@amyeroberts thank you. I have tried to solve all the issues. To summarize, I have finally decided to keep the jax.image.resize layer as it is, even though it results in slightly different output tensors compared to the pytorch model. This is because: 1. the outputs are still close enough to correctly predict the class of the input image and 2. the implementation will be exact once jax updates the resize layer to follow the pytorch conventions in the future. As a result, there was no need to skip any tests. In addition, a PR has been opened on the [hub](https://huggingface.co/facebook/dinov2-base/discussions/12) to add the flax weights. Please let me know if anything is out of order.

**Comment by amyeroberts** [comment]
@MHRDYN7 Great! 

> 2. the implementation will be exact once jax updates the resize layer to follow the pytorch conventions in the future.

Is there a plan for this to happen in the future? 

> In addition, a PR has been opened on the [hub](https://huggingface.co/facebook/dinov2-base/discussions/12) to add the flax weights.

This shouldn't be necessary, all the model frameworks: TF, PyTorch, Flax should be able to load the safetensors file.

**Comment by MHRDYN7** [comment]
> Is there a plan for this to happen in the future?
The [issue](https://github.com/google/jax/issues/15768) from the jax repo mentioned on my first comment, suggests that they did come up with the fix but no steps were taken and also there are no PRs related to the issue. I might just open a PR there if I can; shouldn't be hard to solve. 
>This shouldn't be necessary, all the model frameworks: TF, PyTorch, Flax should be able to load the safetensors file.
It's good to hear that. Indeed, the models don't necessarily need the .msgpack weights. I observed the flax weights on the hub for many models and thought it was a convention to add those.

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:39]: nit - global string should be capitals 

```suggestion
DINOV2_START_DOCSTRING = r"""
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:74]:
```suggestion
DINOV2_INPUTS_DOCSTRING = r"""
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:102]: * not need for the comments if the sizes are dependant on the config values
* comments should go above the relevant code lines to avoid line splitting 

```suggestion
        image_size = self.config.image_size
        patch_size = self.config.patch_size
        image_size = image_size if isinstance(image_size, collections.abc.Iterable) else (image_size, image_size)
        patch_size = patch_size if isinstance(patch_size, collections.abc.Iterable) else (patch_size, patch_size)
        num_patches = (image_size[1] // patch_size[1]) * (image_size[0] // patch_size[0])
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:108]:
```suggestion
            kernel_size=(patch_size[0], patch_size[1]),
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:118]:
```suggestion
        num_channels = pixel_values.shape[-1]
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:768]: no need to truncate 

```suggestion
FLAX_VISION_CLASSIFICATION_DOCSTRING = """
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:785]: Do we want "np" or "jax" here?

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:694]: Same here re "np"

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:610]: Useful to add a comment here to explain what's happening 

```suggestion
        # (batch_size, num_channels, height, width) -> (batch_size, height, width, num_channels)
        pixel_values = jnp.transpose(pixel_values, (0, 2, 3, 1))
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:441]: Why is the condition commented out here?

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:205]:
```suggestion
        height, width = pixel_values.shape[1], pixel_values.shape[2]
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:219]:
```suggestion
        return embeddings
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:211]:
```suggestion
        cls_tokens = jnp.broadcast_to(self.cls_token, (batch_size, 1, self.config.hidden_size))
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:199]:
```suggestion
            (1, num_patches + 1, self.config.hidden_size),
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:195]:
```suggestion
        num_patches = self.patch_embeddings.num_patches
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:191]:
```suggestion
            jax.nn.initializers.variance_scaling(self.config.initializer_range**2, "fan_in", "truncated_normal"),
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:184]:
```suggestion
            jax.nn.initializers.variance_scaling(self.config.initializer_range**2, "fan_in", "truncated_normal"),
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:152]: Why commented out?

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:150]:
```suggestion
    num_patches = hidden_states.shape[1] - 1
    num_positions = position_embeddings.shape[1] - 1
    if num_patches == num_positions and height == width:
        return position_embeddings
    class_pos_embed = position_embeddings[:, 0]
    patch_pos_embed = position_embeddings[:, 1:]
    dim = hidden_states.shape[-1]

    height = height // config.patch_size
    width = width // config.patch_size
    height, width = height + 0.1, width + 0.1

    patch_pos_embed = patch_pos_embed.reshape((1, int(math.sqrt(num_positions)), int(math.sqrt(num_positions)), dim))
    patch_pos_embed = jnp.transpose(patch_pos_embed, (0, 3, 1, 2))

    new_height_ratio = jnp.float32(height / math.sqrt(num_positions))
    new_width_ratio = jnp.float32(width / math.sqrt(num_positions))
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (amyeroberts) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:128]:
```suggestion
        embeddings = self.projection(pixel_values)
        batch_size, _, _, channels = embeddings.shape
        return jnp.reshape(embeddings, (batch_size, -1, channels))
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

Mostly just some nits. Overall LGTM @sanchit-gandhi could you give. quick once-over to confirm flax is OK?

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (MHRDYN7) [reply src/transformers/models/dinov2/modeling_flax_dinov2.py:779]: Should remain "np", I guess, as that's been the case for the previous models

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (MHRDYN7) [reply src/transformers/models/dinov2/modeling_flax_dinov2.py:604]: agree

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (MHRDYN7) [reply src/transformers/models/dinov2/modeling_flax_dinov2.py:433]: The commented out code is from the pytorch model where nn.Identity acts as placeholder (and does nothing to the hidden_states) if drop_path_rate is 0. In flax, however, there is no placeholder function (as far as I know) and that's why this condition has been dealt internally inside FlaxDinov2DropPath. This method is a bit inefficient, but the other flax conversions follow the same design. The comment can be removed.

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (MHRDYN7) [reply src/transformers/models/dinov2/modeling_flax_dinov2.py:146]: It seems that I mistakenly wrote in my last comment that I used jax.image.resize whilst I actually used jax.image.scale_and_translate. Both these functions ultimately call the same helper function internally and therefore both of these could be used for interpolating the tensor. The reason why scale_and_translate() is the better fit is that it allows us to set the scale argument (which is key according to the original Dinov2 repo) while resize() determines the scale on its own. I'll remove the commented out line of code

**Comment by MHRDYN7** [comment]
@amyeroberts thanks a lot for the review. All the tests are passing again.

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (sanchit-gandhi) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:534]: I would copy this from Beit
```suggestion
# Copied from transformers.models.beit.modeling_flax_beit.FlaxBeitPreTrainedModel with Beit-> Dinov2, beit -> dinov2
```

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (sanchit-gandhi) [inline src/transformers/models/dinov2/modeling_flax_dinov2.py:563]: We're missing the rng for the droppath - copying from Beit is going to fix this

File: tests/models/dinov2/test_modeling_flax_dinov2.py
Comment (sanchit-gandhi) [inline tests/models/dinov2/test_modeling_flax_dinov2.py:45]: Can we use batch size 2 for faster testing?
```suggestion
        batch_size=2,
```

File: tests/models/dinov2/test_modeling_flax_dinov2.py
Comment (sanchit-gandhi) [inline tests/models/dinov2/test_modeling_flax_dinov2.py:1]: It'd be super helpful to add `# Copied from` statements in the tests as well!

File: src/transformers/models/dinov2/modeling_flax_dinov2.py
Comment (sanchit-gandhi) [reply src/transformers/models/dinov2/modeling_flax_dinov2.py:771]: Fine to do "np", since we convert "np" arrays to "jnp" arrays before calling the Flax module! (In fact, doing "np" is preferable here, since "jnp" arrays are automatically created on the accelerator device, whereas "np" is always on cpu -> creating your input on cpu and only moving it to accelerator when required is better for async dispatch)

**Review by sanchit-gandhi — COMMENTED** [review]
The PR generally looks in good shape! Well done on handling all of the weight initialisations carefully @MHRDYN7 and porting the new functions over to Flax. 

The main request from my review is using `# Copied from` statements as much as possible. There are many modules / methods that are copied 1-for-1 from existing models in the library. Here, prepending them with a `# Copied from` helps:
1. Keep code sync'd across models
2. The reviewer pinpoint which parts of the code to focus on!

Regarding your PR description: I didn't fully understand what the issue was with the position embedding weights - you've defined them as a standard `self.param`, and the keys look to match those from PyTorch? Let me know if I'm missing something here!

**Comment by MHRDYN7** [comment]
@sanchit-gandhi thanks a lot for the careful review. 

A summary of the updates

- Almost all of your suggestions have been incorporated
- Unable to use `# Copied from`  Beit for the PreTrainedModel class
- Facing some issues trying to use `# Copied from` in the tests, and that's they are kept as they were

**Comment by MHRDYN7** [comment]
> Regarding your PR description: I didn't fully understand what the issue was with the position embedding weights - you've defined them as a standard self.param, and the keys look to match those from PyTorch? Let me know if I'm missing something here!

The position embedding weights can be loaded perfectly. However, these weights are later modified according to the number of patches using F.interpolate (with bicubic mode) in torch. We can replicate this behavior with jax.image.scale_and_translate (or with image.resize), but it [seems](https://github.com/google/jax/issues/15768) that this function is slight different from torch interpolate only in case of the bicubic mode, resulting slightly different output hidden_states.

Please Note:
the default config is image_size = 224, patch_size = 16 (and interpolation of the pos_embeds are not needed in this case); 
but the pretrained weights config on hub is image_size = 518, patch_size = 14

**Comment by amyeroberts** [comment]
Thanks for the detailed explanation and iterating with us @MHRDYN7! @sanchit-gandhi is off at the moment, but I can see you've addressed his comments, so I think we're OK to merge without his second review.

Final step is running the slow tests for the model before merge. Could you push an empty commit with the message `[run_slow] dinov2`?

**Comment by MHRDYN7** [comment]
@amyeroberts, I've pushed the required commit. Now I guess, it requires your approval for running the slow tests

**Comment by MHRDYN7** [comment]
@amyeroberts slow tests passed! Ready to be merged

**Review by amyeroberts — APPROVED** [review]
Great piece of work - thanks for adding!

**Pull request merged**
